File: typescript/scripts/createReport.ts
/**
 * Interactive CLI script: create and run a Riverscapes Report end-to-end.
 *
 * Workflow:
 * 1. Authenticate — a browser tab opens for you to log in (OAuth 2.0 PKCE).
 * 2. Pick a report type — only types with the PICK input tool are shown.
 * 3. Name the report.
 * 4. Select a picker layer (if the report type defines any).
 * 5. Enter a picker item ID.
 * 6. Choose a unit system (if the report type supports multiple).
 * 7. Create the report.
 * 8. Attach the picker option.
 * 9. Start the report.
 * 10. Poll for completion.
 * 11. Print a link to the finished report.
 *
 * Usage:
 *   npx tsx scripts/createReport.ts staging
 *   npx tsx scripts/createReport.ts production
 */

import { select, input } from '@inquirer/prompts'
import { ReportsAPI, type RSReportType } from '../src/index.js'

function layerLabel(layerId: string): string {
  return layerId.replace(/_/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase())
}

async function main(): Promise<void> {
  const stage = process.argv[2]
  if (!stage || !['staging', 'production', 'local'].includes(stage)) {
    console.error('Usage: npx tsx scripts/createReport.ts <staging|production|local>')
    process.exit(1)
  }

  const api = new ReportsAPI({ stage })
  await api.open()

  try {
    console.log(`\n=== Creating report on ${stage.toUpperCase()} ===\n`)

    // ---- List report types and pick one ----
    const reportTypes = await api.listReportTypes()

    // Filter to types with PICK tool
    const filtered = reportTypes.filter((rt) => {
      const tools = (rt.parameters as Record<string, unknown> | undefined)?.['tools'] as string[] | undefined
      return tools?.includes('PICK')
    })

    const reportType: RSReportType = await select({
      message: 'Select report type:',
      choices: filtered.map((rt) => ({
        name: `${rt.name}  (v${rt.version})`,
        value: rt,
      })),
    })

    // ---- Enter a report name ----
    const defaultName = `${reportType.name} - ${new Date().toISOString().replace('T', ' ').slice(0, 19)}`
    const name = await input({
      message: 'Report name:',
      default: defaultName,
      validate: (v) => (v.trim() ? true : 'Name cannot be empty'),
    })

    // ---- Pick a picker layer ----
    const params = (reportType.parameters ?? {}) as Record<string, unknown>
    const validLayers = (params['validPickerLayers'] ?? []) as string[]

    let pickerLayer: string | null = null
    if (validLayers.length > 0) {
      pickerLayer = await select({
        message: 'Select picker layer:',
        choices: validLayers.map((layer) => ({
          name: layerLabel(layer),
          value: layer,
        })),
      })
    }

    // ---- Enter a picker item ID ----
    let pickerId: string | null = null
    if (pickerLayer) {
      pickerId = await input({
        message: `Enter ID for ${layerLabel(pickerLayer)}:`,
        default: '1302020710',
      })
    }

    // ---- Choose a unit system ----
    const parameters: Record<string, unknown> = {}
    const validUnitSystems = (params['validUnitSystems'] ?? []) as string[]
    if (validUnitSystems.length > 0) {
      parameters['units'] = await select({
        message: 'Select unit system:',
        choices: validUnitSystems.map((u) => ({
          name: u.toUpperCase(),
          value: u,
        })),
      })
    }

    // ---- Create report ----
    let report = await api.createReport({
      name: name.trim(),
      reportTypeId: reportType.id!,
      parameters,
    })

    console.log(`\x1b[32mReport created: ${report.id}\x1b[0m`)
    console.log(`  Name:   ${report.name}`)
    console.log(`  Status: ${report.status}`)
    console.log(`  Type:   ${report.reportType?.name ?? reportType.id}`)

    // ---- Attach picker option ----
    if (pickerLayer && pickerId) {
      console.log(`Attaching picker option: ${layerLabel(pickerLayer)} = ${pickerId}`)
      await api.attachPickerOption(report.id!, pickerLayer, pickerId)
      console.log(`\x1b[32mPicker option attached.\x1b[0m`)
    }

    // ---- Start report ----
    console.log('Starting report...')
    report = await api.startReport(report.id!)
    console.log(`\x1b[36mReport started (status: ${report.status})\x1b[0m`)

    // ---- Poll for completion ----
    console.log('Polling for completion every 10 seconds...')
    report = await api.pollReport(report.id!, 10)

    console.log()
    if (report.isComplete()) {
      console.log(`\x1b[32mReport COMPLETE!\x1b[0m`)
      const frontendUrls: Record<string, string> = {
        production: 'https://reports.riverscapes.net',
        staging: 'https://staging.reports.riverscapes.net',
      }
      const base = frontendUrls[stage]
      if (base && report.createdById) {
        console.log(`\x1b[32m  View report: ${base}/reports/${report.createdById}/${report.id}/report.html\x1b[0m`)
      } else if (base) {
        console.log(`\x1b[32m  View your reports: ${base}/my\x1b[0m`)
      }
    } else {
      console.error(`\x1b[31mReport ended with status: ${report.status}\x1b[0m`)
      if (report.statusMessage) {
        console.error(`  Message: ${report.statusMessage}`)
      }
    }
    console.log(`  Report ID: ${report.id}`)
    console.log()
  } finally {
    api.close()
  }
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})

File: python/scripts/create_report.py
"""Interactive CLI script: create and run a Riverscapes Report end-to-end.

This script walks you through every step needed to produce a report, asking
questions at each stage so you do not need to write any code.

Run this script using the launch menu under "🐍 Python: Run/Debug Current File".  You will
be prompted to select which API stage to target (staging or production), then you will be 
guided through the workflow described below.

Workflow
--------
1. **Authenticate** \u2014 a browser tab opens for you to log in with your
   Riverscapes account (OAuth 2.0 PKCE flow).  The token is cached for the
   duration of the script.
2. **Pick a report type** \u2014 only types that support the ``PICK`` input
   tool are shown (i.e. those that need you to select an existing geographic
   feature rather than draw or upload one).
3. **Name the report** \u2014 a default name with a timestamp is pre-filled.
4. **Select a picker layer** \u2014 if the report type declares ``validPickerLayers``,
   you choose which layer to select from (e.g. ``huc``, ``catchment``).
5. **Enter a picker item ID** \u2014 the ID of the specific feature in the chosen
   layer (e.g. a HUC-12 code like ``1302020710``).
6. **Choose a unit system** \u2014 if the report type supports multiple unit systems
   (e.g. ``SI`` / ``imperial``), you pick one.
7. **Create the report** \u2014 a ``CREATED`` record is registered in the API and
   an ID is returned.
8. **Attach picker option** \u2014 the chosen layer + ID is linked to the report.
9. **Start the report** \u2014 the report is submitted to the processing queue.
10. **Poll for completion** \u2014 the script checks the status every 10 seconds
    and prints progress until the report reaches a terminal state.
11. **Print a link** \u2014 on success a direct URL to the finished report is shown.

Usage
-----
::

    python scripts/create_report.py staging
    python scripts/create_report.py production

Arguments
---------
stage : {'staging', 'production', 'local'}
    Which API environment to target.  Use ``staging`` while developing;
    use ``production`` for real work.
"""
from datetime import datetime
import argparse
import questionary
from rsxml import Logger
from termcolor import colored
from pyreports import ReportsAPI

log = Logger('Create Report')


def layer_label(layer_id: str) -> str:
    """Convert a picker layer ID to a human-readable label."""
    return layer_id.replace('_', ' ').title()


def main():
    """Entry point for the interactive report-creation CLI.

    Parses the ``stage`` positional argument, authenticates with the API,
    and walks the user through a series of ``questionary`` prompts to
    configure and launch a report.  Blocks until the report reaches a
    terminal state, then prints a summary and (on success) a direct URL
    to view the report on the Riverscapes Reports web frontend.
    """
    parser = argparse.ArgumentParser(description="Create and run a report interactively.")
    parser.add_argument('stage', choices=['staging', 'production', 'local'], help="API stage")
    args = parser.parse_args()

    with ReportsAPI(stage=args.stage) as api:
        log.title(f"🚀 Creating report on {args.stage.upper()}")

        ################## API CALL: List report types and pick one ##################
        report_types = api.list_report_types()

        # Filter out anything that doesn't have PickToolsEnum.PICK inside parameters.tools.
        # This tool currently doesn't support DRAW or UPLOAD (but it couls)
        report_types_filtered = [x for x in report_types if 'PICK' in (x.parameters or {}).get('tools', [])]
        report_type = questionary.select(
            "Select report type:",
            choices=[
                questionary.Choice(f"{rt.name}  (v{rt.version})", value=rt)
                for rt in report_types_filtered
            ],
        ).ask()
        if report_type is None:
            return

        # Enter a report name, set the default to be just the report type name with a timestamp, but allow the user to change it.
        name = questionary.text(
            "Report name:",
            validate=lambda v: "Name cannot be empty" if not v.strip() else True,
            default=f"{report_type.name} - {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}"
        ).ask()
        if name is None:
            return

        # Pick a picker layer (if the report type defines any)
        picker_layer = None
        valid_layers = (report_type.parameters or {}).get('validPickerLayers', [])
        if valid_layers:
            picker_layer = questionary.select("Select picker layer:", choices=[
                questionary.Choice(layer_label(layer), value=layer)
                for layer in valid_layers
            ],
            ).ask()
            if picker_layer is None:
                return

        # Now we have to enter an ID as a string that corresponds to the id of the object in the picker layer
        picker_id = None
        if picker_layer:
            picker_id = questionary.text(f"Enter ID for {layer_label(picker_layer)}:", default='1302020710').ask()
            if picker_id is None:
                return

        parameters = {}
        # If validUnitSystems is defined inside tool type parmaters then we need to choose from the opsions       validUnitSystems: ['SI', 'imperial'], // Remove this property if UNITS is not a valid parameter for this report ORDER IS PRESENTATION ORDER. FIRST ITEM IS THE DEFAULT
        valid_unit_systems = (report_type.parameters or {}).get('validUnitSystems', [])
        if valid_unit_systems:
            parameters['units'] = questionary.select("Select unit system:", choices=[
                questionary.Choice(unit_system.upper(), value=unit_system)
                for unit_system in valid_unit_systems
            ]).ask()

        #################### API CALL: Create report ##################
        # We create an empty report in order to have an id we can attach our input geojson to
        report = api.create_report(
            name=name.strip(),
            report_type_id=report_type.id,
            parameters=parameters,
        )
        log.info(colored(f"✅ Report created: {report.id}", 'green'))
        log.info(f"  📛 Name:   {report.name}")
        log.info(f"  📊 Status: {report.status}")
        log.info(f"  📋 Type:   {report.report_type.name if report.report_type else report_type.id}")

        # Attach the picker option (if a picker layer was selected)
        if picker_layer and picker_id:
            log.info(f"📎 Attaching picker option: {layer_label(picker_layer)} = {picker_id}")

            #################### API CALL: Attach picker option to report ##################
            api.attach_picker_option(report.id, picker_layer, picker_id)

            log.info(colored("✅ Picker option attached.", 'green'))

        # Start the report
        log.info(colored("▶️  Starting report...", 'cyan'))

        ##################### API CALL: Start report ##################
        report = api.start_report(report.id)
        log.info(colored(f"🔄 Report started (status: {report.status})", 'cyan'))

        # Poll every 10 seconds until a terminal state is reached
        log.info(colored("⏳ Polling for completion every 10 seconds...", 'cyan'))

        ##################### API CALL: Poll report ##################
        report = api.poll_report(report.id, interval=10)

        print()
        if report.is_complete():
            log.info(colored("🎉 Report COMPLETE!", 'green'))
            frontend_urls = {
                'production': 'https://reports.riverscapes.net',
                'staging': 'https://staging.reports.riverscapes.net',
            }
            base = frontend_urls.get(args.stage)
            if base and report.created_by_id:
                print(colored(f"  🔗 View report: {base}/reports/{report.created_by_id}/{report.id}/report.html", 'green'))
            elif base:
                print(colored(f"  🔗 View your reports: {base}/my", 'green'))
        else:
            log.error(colored(f"❌ Report ended with status: {report.status}", 'red'))
            if report.status_message:
                log.error(f"  💬 Message: {report.status_message}")
        print(f"  Report ID: {report.id}")
        print()


if __name__ == '__main__':
    main()
